std/arith.asm

```diff
@@ -17,18 +24,184 @@ machine Arith(CLK32_31, operation_id){
+    pol commit y1__0, y1__1, y1__2, y1__3, y1__4, y1__5, y1__6, y1__7, y1__8, y1__9, y1__10, y1__11, y1__12, y1__13, y1__14, y1__15;
+    let y1 = [y1__0, y1__1, y1__2, y1__3, y1__4, y1__5, y1__6, y1__7, y1__8, y1__9, y1__10, y1__11, y1__12, y1__13, y1__14, y1__15];
+    pol commit x2__0, x2__1, x2__2, x2__3, x2__4, x2__5, x2__6, x2__7, x2__8, x2__9, x2__10, x2__11, x2__12, x2__13, x2__14, x2__15;
+    let x2 = [x2__0, x2__1, x2__2, x2__3, x2__4, x2__5, x2__6, x2__7, x2__8, x2__9, x2__10, x2__11, x2__12, x2__13, x2__14, x2__15];
+    pol commit y2__0, y2__1, y2__2, y2__3, y2__4, y2__5, y2__6, y2__7, y2__8, y2__9, y2__10, y2__11, y2__12, y2__13, y2__14, y2__15;
+    let y2 = [y2__0, y2__1, y2__2, y2__3, y2__4, y2__5, y2__6, y2__7, y2__8, y2__9, y2__10, y2__11, y2__12, y2__13, y2__14, y2__15];
+    pol commit x3__0, x3__1, x3__2, x3__3, x3__4, x3__5, x3__6, x3__7, x3__8, x3__9, x3__10, x3__11, x3__12, x3__13, x3__14, x3__15;
+    let x3 = [x3__0, x3__1, x3__2, x3__3, x3__4, x3__5, x3__6, x3__7, x3__8, x3__9, x3__10, x3__11, x3__12, x3__13, x3__14, x3__15];
+    pol commit y3__0, y3__1, y3__2, y3__3, y3__4, y3__5, y3__6, y3__7, y3__8, y3__9, y3__10, y3__11, y3__12, y3__13, y3__14, y3__15;
+    let y3 = [y3__0, y3__1, y3__2, y3__3, y3__4, y3__5, y3__6, y3__7, y3__8, y3__9, y3__10, y3__11, y3__12, y3__13, y3__14, y3__15];
+
+    // Selects the ith limb of x (little endian)
+    // Note that the most significant limb can be up to 32 bits; all others are 16 bits.
+    let select_limb = |x, i| fe((x >> (i * 16)) & if i < 15 { 0xffff } else { 0xffffffff });
+
+    let s_for_eq1 = |x1, y1, x2, y2| div(sub(y2, y1), sub(x2, x1));
+    let s_for_eq2 = |x1, y1| div(mul(3, mul(x1, x1)), mul(2, y1));
+
+    // Adding secp_modulus to make sure that that all numbers are positive when % is applied to it.
+    let compute_x3_int = |x1, x2, s| (s * s - x1 - x2 + 2 * secp_modulus) % secp_modulus;
+    let compute_y3_int = |x1, y1, x3, s| (s * ((x1 - x3) + secp_modulus) - y1 + secp_modulus) % secp_modulus;
+
+    // Compute quotients for the various equation.
+    // Note that we add 2**258 to it, to move it from the (-2**258, 2**258) to the (0, 2**259) range, so it can
+    // be represented as an unsigned 272-bit integer.
+    // See the comment for `product_with_p` below.
+    let compute_q0_for_eq1 = |x1, y1, x2, y2, s| (-(s * x2 - s * x1 - y2 + y1) / secp_modulus + (1 << 258));
+    let compute_q0_for_eq2 = |x1, y1, s| (-(2 * s * y1 - 3 * x1 * x1) / secp_modulus + (1 << 258));
+    let compute_q1 = |x1, x2, x3, s| (-(s * s - x1 - x2 - x3) / secp_modulus + (1 << 258));
+    let compute_q2 = |x1, y1, x3, y3, s| (-(s * x1 - s * x3 - y1 - y3) / secp_modulus + (1 << 258));
+ 
+    let limbs_to_int: expr[] -> int = |limbs| array::sum(array::map_enumerated(limbs, |i, limb| int(eval(limb)) << (i * 16)));
+
+    let x1_int: -> int = || limbs_to_int(x1);
```

Comment:
Ok this is a bit weird because in a pure functional sense, this should be a constant. Maybe we can at some point also mark functions that need to be evaluated in a query context.

Reply:
What is also not very clear to me is that the way this ends up being used is:
```
col witness s_0(i) query ("hint", fe(select_limb(s_hint(), 0)));
```

So the call to `s_hint()` is evaluated at witgen time (because otherwise it would hit `eval()`, which would fail?). Maybe it would be good to make that explicit somehow, e.g.:
 ```
col witness s_0: witness = |i| ("hint", fe(select_limb(s_hint(), 0)));
```

Then witness and fixed columns would be specified the same, but depending on the type the callback gets executed in different phases? I think that would also give us arrays of witness columns with hints, just like we already have arrays of fixed column?